java/src/main/native/src/row_conversion.cu

```diff
@@ -594,134 +594,106 @@ __global__ void copy_to_rows(const size_type num_rows, const size_type num_colum
+      auto const col_size = col_sizes[absolute_col];
+      auto const col_offset = col_offsets[absolute_col];
+      auto const relative_col_offset = col_offset - starting_column_offset;
+
+      auto const shared_offset = relative_row * tile_row_size + relative_col_offset;
+      auto const input_src = input_data[absolute_col] + col_size * absolute_row;
+
+      // copy the element from global memory
+      switch (col_size) {
+        case 2: {
+          const int16_t *short_col_input = reinterpret_cast<const int16_t *>(input_src);
+          *reinterpret_cast<int16_t *>(&shared_data[shared_offset]) = *short_col_input;
+          break;
         }
-        auto const absolute_row = relative_row + fetch_tile.start_row;
-        auto const col_size = col_sizes[absolute_col];
-        auto const col_offset = col_offsets[absolute_col];
-        auto const relative_col_offset = col_offset - starting_column_offset;
-
-        auto const shared_offset = relative_row * fetch_tile_row_size + relative_col_offset;
-        auto const input_src = input_data[absolute_col] + col_size * absolute_row;
-
-        // copy the element from global memory
-        switch (col_size) {
-          case 2:
-            MEMCPY(&shared_buffer_base[shared_offset], input_src, aligned_size_t<2>(col_size),
-                   fetch_barrier);
-            break;
-          case 4:
-            MEMCPY(&shared_buffer_base[shared_offset], input_src, aligned_size_t<4>(col_size),
-                   fetch_barrier);
-            break;
-          case 8:
-            MEMCPY(&shared_buffer_base[shared_offset], input_src, aligned_size_t<8>(col_size),
-                   fetch_barrier);
-            break;
-          default:
-            MEMCPY(&shared_buffer_base[shared_offset], input_src, col_size, fetch_barrier);
-            break;
+        case 4: {
+          const int32_t *int_col_input = reinterpret_cast<const int32_t *>(input_src);
+          *reinterpret_cast<int32_t *>(&shared_data[shared_offset]) = *int_col_input;
+          break;
+        }
+        case 8: {
+          const int64_t *long_col_input = reinterpret_cast<const int64_t *>(input_src);
+          *reinterpret_cast<int64_t *>(&shared_data[shared_offset]) = *long_col_input;
+          break;
+        }
+        case 1: shared_data[shared_offset] = *input_src; break;
+        default: {
+          for (int i = 0; i < col_size; ++i) {
+            shared_data[shared_offset] = *input_src;
+          }
+          break;
         }
       }
     }
+  }
+
+  auto const tile_output_buffer = output_data[tile.batch_number];
+  auto const row_batch_start = tile.batch_number == 0 ? 0 : batch_row_boundaries[tile.batch_number];
 
 #ifdef ASYNC_MEMCPY_SUPPORTED
-    auto &processing_barrier = tile_barrier[processing_index % NUM_TILES_PER_KERNEL_LOADED];
-    processing_barrier.arrive_and_wait();
+  tile_barrier.arrive_and_wait();
 #else
-    group.sync();
+  group.sync();
 #endif // ASYNC_MEMCPY_SUPPORTED
```

Comment:
Calling out what we observed here: that we'll need to unconditionally `group.sync()`, and not wait on the `tile_barrier`.